reference/solver/lobpcg_kernels.cpp

```diff
@@ -0,0 +1,108 @@
+// SPDX-FileCopyrightText: 2017 - 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "core/solver/lobpcg_kernels.hpp"
+
+#include <ginkgo/core/base/types.hpp>
+
+#include "reference/base/lapack_bindings.hpp"
+
+#if GKO_HAVE_LAPACK
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The LOBPCG solver namespace.
+ *
+ * @ingroup lobpcg
+ */
+namespace lobpcg {
+
+
+using gko::kernels::lobpcg::workspace_mode;
+
+template <typename ValueType>
+void symm_generalized_eig(std::shared_ptr<const ReferenceExecutor> exec,
+                          const workspace_mode alloc,
+                          matrix::Dense<ValueType>* a,
+                          matrix::Dense<ValueType>* b,
+                          array<remove_complex<ValueType>>* e_vals,
+                          array<char>* workspace)
+{
+    const char job = LAPACK_EIG_VECTOR;
+    const char uplo = LAPACK_FILL_LOWER;
+    const int n = static_cast<int>(a->get_size()[1]);
+    const int lda = static_cast<int>(a->get_stride());
+    const int ldb = static_cast<int>(b->get_stride());
+    const int itype = 1;
+    if constexpr (!gko::is_complex_s<ValueType>::value) {
+        // Even if the workspace is already allocated, we need to know where to
+        // se set the pointers for the individual workspaces of LAPACK
+        int fp_buffer_num_elems;
+        int int_buffer_num_elems;
+        array<int> tmp_iwork(exec, 1);
+        lapack::sygvd_buffersizes(
+            &itype, &job, &uplo, &n, a->get_values(), &lda, b->get_values(),
+            &ldb, e_vals->get_data(), (ValueType*)workspace->get_data(),
+            &fp_buffer_num_elems, tmp_iwork.get_data(), &int_buffer_num_elems);
+        if (alloc == workspace_mode::allocate) {
```

Comment:
we only need to check whether the workspace is enough for this operation. If not, we reallocate enough workspace for that. Currently, I think we do not handle the case that user does not let us allocate anything.

Reply:
I should add, this applies to the cuSOLVER/rocSOLVER case specifically (that we can avoid calling the query routine sometimes).

Reply:
The `workspace_mode` has been removed in 24e006d8a3419f4363e899ed896447de5107698d